File: templates/typescript/discovery/Libp2pDiscovery.ts
/**
 * Libp2pDiscovery — fully P2P discovery backend for Sentrix.
 *
 * Architecture:
 *   Transport   : QUIC (via @chainsafe/libp2p-quic) — no TCP listener opened
 *   Routing     : Kademlia DHT (custom /sentrix/kad/1.0.0 protocol)
 *   Local LAN   : mDNS (optional, default on)
 *   NAT         : DCUtR hole punching + circuit-relay-v2 fallback
 *   Identity    : secp256k1 keypair from ANR — same key, same PeerId
 *
 * Capability discovery uses DHT provider records:
 *   SHA256("sentrix:cap:<capability>") → CIDv1 → dht.provide() / findProviders()
 *
 * Full DiscoveryEntry is stored as a signed JSON envelope:
 *   SHA256("sentrix:anr:<agentId>") → key → dht.put() / get()
 *
 * A reverse PeerId→agentId mapping is also stored so findProviders() results
 * can be resolved to full entries.
 *
 * Usage:
 *   const discovery = await Libp2pDiscovery.create({ privateKey: myKey });
 *   await discovery.register(entry);
 *   const peers = await discovery.query('web_search');
 *   await discovery.stop();
 *
 * @see docs/libp2p.md for full architecture documentation.
 */

import { createLibp2p, type Libp2p } from 'libp2p';
import { quic }                       from '@chainsafe/libp2p-quic';
import { circuitRelayTransport }      from '@libp2p/circuit-relay-v2';
import { identify }                   from '@libp2p/identify';
import { kadDHT, type KadDHT }        from '@libp2p/kad-dht';
import { mdns }                       from '@libp2p/mdns';
import { dcutr }                      from '@libp2p/dcutr';
import { bootstrap }                  from '@libp2p/bootstrap';
import { toString, fromString }       from 'uint8arrays';
import type { PeerId }                from '@libp2p/interface';

import { IAgentDiscovery, DiscoveryEntry } from '../interfaces/IAgentDiscovery';
import { capabilityCid, anrDhtKey, pidDhtKey } from './libp2p/DhtKeys';
import { encodeEnvelope, decodeEnvelope }       from './libp2p/EntryEnvelope';
import { peerIdFromAnrKey, publicKeyFromAnrKey } from './libp2p/PeerIdFromAnr';
import { keys }                                  from '@libp2p/crypto';

// ── Types ────────────────────────────────────────────────────────────────────

export interface Libp2pDiscoveryConfig {
  /**
   * 32-byte raw secp256k1 private key — the same key used to sign ANR records.
   * If omitted, an ephemeral key is generated (not recommended for production).
   */
  privateKey?: Uint8Array;

  /**
   * Multiaddrs this node will listen on.
   * Default: ['/ip4/0.0.0.0/udp/0/quic-v1']  (random UDP port)
   */
  listenAddresses?: string[];

  /**
   * Multiaddrs of known bootstrap peers (format: /ip4/.../udp/.../quic-v1/p2p/...)
   * Also reads from SENTRIX_BOOTSTRAP_PEERS env var (comma-separated).
   */
  bootstrapPeers?: string[];

  /**
   * How often to re-publish DHT records to keep them alive (milliseconds).
   * Default: 30_000 (30 seconds)
   */
  heartbeatIntervalMs?: number;

  /**
   * Whether to enable mDNS for local network discovery.
   * Default: true
   * Disable in environments where LAN broadcast is undesirable.
   */
  enableMdns?: boolean;

  /**
   * Set to true to operate in DHT client mode — the node participates in
   * discovery but does not store records for others.  Reduces resource usage
   * on constrained agents.
   * Default: false
   */
  dhtClientMode?: boolean;
}

interface LocalEntry {
  entry: DiscoveryEntry;
  seq:   number;
  publicKey: Uint8Array;
}

// ── Sentinel bootstrap peers for the public Sentrix network ──────────────────
// Replace these with real multiaddrs when production bootstrap nodes are deployed.
const SENTRIX_BOOTSTRAP_PEERS: string[] = [
  // '/ip4/bootstrap1.sentrix.io/udp/4001/quic-v1/p2p/12D3KooWXXX...',
];

// ── Libp2pDiscovery ───────────────────────────────────────────────────────────

export class Libp2pDiscovery implements IAgentDiscovery {
  private readonly node:    Libp2p;
  private readonly dht:     KadDHT;
  private readonly privKey: Uint8Array;
  private readonly pubKey:  Uint8Array;

  /** Locally registered agents: agentId → entry + seq + public key */
  private localEntries = new Map<string, LocalEntry>();

  /** Background heartbeat timers: agentId → NodeJS.Timer */
  private heartbeatTimers = new Map<string, ReturnType<typeof setInterval>>();

  /** Short-lived query cache: capability → { entries, expiresAt } */
  private queryCache = new Map<string, { entries: DiscoveryEntry[]; expiresAt: number }>();

  private constructor(node: Libp2p, dht: KadDHT, privKey: Uint8Array, pubKey: Uint8Array) {
    this.node    = node;
    this.dht     = dht;
    this.privKey = privKey;
    this.pubKey  = pubKey;
  }

  // ── Factory ─────────────────────────────────────────────────────────────────

  /**
   * Create and start a Libp2pDiscovery node.
   * This is an async factory because libp2p.start() is asynchronous.
   */
  static async create(config: Libp2pDiscoveryConfig = {}): Promise<Libp2pDiscovery> {
    // ── Keypair ──────────────────────────────────────────────────────────────
    let privKey: Uint8Array;
    let libp2pPrivKey: Awaited<ReturnType<typeof keys.generateKeyPairFromSeed>>;

    if (config.privateKey) {
      privKey      = config.privateKey;
      libp2pPrivKey = await keys.generateKeyPairFromSeed('secp256k1', privKey);
    } else {
      // Ephemeral key — generates random secp256k1 keypair
      libp2pPrivKey = await keys.generateKeyPair('secp256k1');
      privKey        = (libp2pPrivKey as any).raw ?? new Uint8Array(32); // fallback
      console.warn('[Libp2pDiscovery] Using ephemeral identity — set privateKey for production');
    }
    const pubKey = libp2pPrivKey.publicKey.raw;

    // ── Bootstrap peers ──────────────────────────────────────────────────────
    const envPeers    = process.env['SENTRIX_BOOTSTRAP_PEERS']?.split(',').filter(Boolean) ?? [];
    const allBootstrap = [
      ...(config.bootstrapPeers ?? []),
      ...envPeers,
      ...SENTRIX_BOOTSTRAP_PEERS,
    ];

    // ── libp2p node ──────────────────────────────────────────────────────────
    const peerDiscovery = [
      ...(config.enableMdns !== false ? [mdns()] : []),
      ...(allBootstrap.length > 0    ? [bootstrap({ list: allBootstrap })] : []),
    ];

    const node = await createLibp2p({
      privateKey: libp2pPrivKey,
      addresses: {
        listen: config.listenAddresses ?? ['/ip4/0.0.0.0/udp/0/quic-v1'],
      },
      transports: [
        quic(),
        // Circuit relay transport for NAT fallback (outbound connections only)
        circuitRelayTransport({ discoverRelays: 1 }),
      ],
      // QUIC handles TLS natively — no separate connectionEncrypter needed
      peerDiscovery,
      services: {
        identify: identify(),
        dht: kadDHT({
          protocol:   '/sentrix/kad/1.0.0',  // isolated from public IPFS DHT
          clientMode: config.dhtClientMode ?? false,
        }),
        dcutr:  dcutr(),
      },
    });

    await node.start();

    const dht = node.services['dht'] as unknown as KadDHT;

    console.log(
      `[Libp2pDiscovery] Started — PeerId: ${node.peerId}  ` +
      `Addrs: ${node.getMultiaddrs().map(String).join(', ') || '(none yet)'}`,
    );

    return new Libp2pDiscovery(node, dht, privKey, pubKey);
  }

  /**
   * Return the local node's PeerId and first multiaddr, or null if not started.
   * Used by ExampleAgent to populate the `multiaddr` field in ANR records.
   */
  getNodeInfo(): { peerId: string; multiaddr: string } | null {
    const peerId = this.node.peerId.toString();
    const addrs  = this.node.getMultiaddrs();
    if (!peerId) return null;
    return { peerId, multiaddr: addrs.length > 0 ? addrs[0].toString() : '' };
  }

  /** Gracefully stop the libp2p node and clear all timers. */
  async stop(): Promise<void> {
    for (const timer of this.heartbeatTimers.values()) clearInterval(timer);
    this.heartbeatTimers.clear();
    await this.node.stop();
    console.log('[Libp2pDiscovery] Stopped');
  }

  // ── IAgentDiscovery ──────────────────────────────────────────────────────────

  async register(entry: DiscoveryEntry): Promise<void> {
    const seq = (this.localEntries.get(entry.agentId)?.seq ?? 0) + 1;

    // Persist locally first so we can respond even if DHT is not yet connected
    this.localEntries.set(entry.agentId, { entry, seq, publicKey: this.pubKey });

    // Publish to DHT (fire-and-forget with error logging)
    this.publishToDht(entry, seq).catch(err =>
      console.warn(`[Libp2pDiscovery] DHT publish pending (${entry.agentId}): ${err.message}`)
    );

    // Start heartbeat refresh timer
    this.startHeartbeat(entry.agentId, entry, seq);

    console.log(`[Libp2pDiscovery] Registered: ${entry.agentId}  caps=${entry.capabilities}`);
  }

  async unregister(agentId: string): Promise<void> {
    const timer = this.heartbeatTimers.get(agentId);
    if (timer) { clearInterval(timer); this.heartbeatTimers.delete(agentId); }
    this.localEntries.delete(agentId);
    // DHT provider records expire naturally; we cannot actively revoke them.
    // Mark the value record as unhealthy so peers stop using it.
    console.log(`[Libp2pDiscovery] Unregistered: ${agentId}`);
  }

  async query(capability: string): Promise<DiscoveryEntry[]> {
    // 1. Check short-lived cache (10s TTL)
    const cached = this.queryCache.get(capability);
    if (cached && cached.expiresAt > Date.now()) return cached.entries;

    // 2. Check local entries first (zero network latency)
    const local = [...this.localEntries.values()]
      .filter(e => e.entry.capabilities.includes(capability) && e.entry.health.status !== 'unhealthy')
      .map(e => e.entry);

    // 3. Query DHT for remote providers
    const remote: DiscoveryEntry[] = [];
    try {
      const capCid = await capabilityCid(capability);

      for await (const event of this.dht.findProviders(capCid, { signal: AbortSignal.timeout(5_000) })) {
        if ((event as any).name !== 'PROVIDER') continue;
        const providerId: PeerId = (event as any).peer?.id;
        if (!providerId) continue;
        if (providerId.equals(this.node.peerId)) continue; // skip self

        const entry = await this.fetchEntryForPeer(providerId);
        if (entry && entry.health.status !== 'unhealthy') remote.push(entry);
      }
    } catch (err: any) {
      console.warn(`[Libp2pDiscovery] DHT findProviders failed for '${capability}': ${err.message}`);
    }

    // 4. Merge, deduplicate by agentId
    const seen = new Set<string>();
    const all: DiscoveryEntry[] = [];
    for (const e of [...local, ...remote]) {
      if (!seen.has(e.agentId)) { seen.add(e.agentId); all.push(e); }
    }

    // 5. Cache for 10 seconds
    this.queryCache.set(capability, { entries: all, expiresAt: Date.now() + 10_000 });
    return all;
  }

  async listAll(): Promise<DiscoveryEntry[]> {
    // Local entries are authoritative; remote is best-effort
    const local = [...this.localEntries.values()].map(e => e.entry);
    return local;
    // Note: a full DHT walk for remote agents is expensive and eventually-consistent.
    // In production, use a dedicated gossip channel or a list built from accumulated
    // query results. See docs/libp2p.md § listAll.
  }

  async heartbeat(agentId: string): Promise<void> {
    const local = this.localEntries.get(agentId);
    if (!local) {
      console.warn(`[Libp2pDiscovery] heartbeat called for unknown agent: ${agentId}`);
      return;
    }
    const newSeq   = local.seq + 1;
    const newEntry = {
      ...local.entry,
      health: {
        ...local.entry.health,
        status: 'healthy' as const,
        lastHeartbeat: new Date().toISOString(),
      },
    };
    this.localEntries.set(agentId, { ...local, entry: newEntry, seq: newSeq });
    await this.publishToDht(newEntry, newSeq);
  }

  // ── Private helpers ──────────────────────────────────────────────────────────

  /** Publish a DiscoveryEntry + all its capability keys to the DHT. */
  private async publishToDht(entry: DiscoveryEntry, seq: number): Promise<void> {
    const encoded   = encodeEnvelope(entry, seq, this.privKey);
    const valueKey  = await anrDhtKey(entry.agentId);
    const pidKey    = pidDhtKey(this.node.peerId.toString());

    await Promise.all([
      // (a) Store the full DiscoveryEntry envelope
      this.dht.put(valueKey, encoded),

      // (b) Store the PeerId → agentId reverse mapping
      this.dht.put(pidKey, new TextEncoder().encode(entry.agentId)),

      // (c) Announce as provider for every capability
      ...entry.capabilities.map(async cap => {
        const cid = await capabilityCid(cap);
        await this.dht.provide(cid);
      }),
    ]);
  }

  /** Fetch a DiscoveryEntry for a remote peer via the DHT. */
  private async fetchEntryForPeer(peerId: PeerId): Promise<DiscoveryEntry | null> {
    try {
      // Step 1: resolve PeerId → agentId
      const pidKey   = pidDhtKey(peerId.toString());
      const agentIdRaw = await this.dhtGet(pidKey);
      if (!agentIdRaw) return null;

      const agentId  = new TextDecoder().decode(agentIdRaw);

      // Step 2: resolve agentId → DiscoveryEntry envelope
      const valueKey = await anrDhtKey(agentId);
      const raw      = await this.dhtGet(valueKey);
      if (!raw) return null;

      const decoded  = decodeEnvelope(raw);
      return decoded?.entry ?? null;
    } catch {
      return null;
    }
  }

  /** Wraps dht.get() to consume the async iterator and return the first value. */
  private async dhtGet(key: Uint8Array): Promise<Uint8Array | null> {
    for await (const event of this.dht.get(key, { signal: AbortSignal.timeout(3_000) })) {
      if ((event as any).name === 'VALUE') {
        return (event as any).value as Uint8Array;
      }
    }
    return null;
  }

  /** Start a background timer that re-publishes DHT records at heartbeatIntervalMs. */
  private startHeartbeat(
    agentId: string,
    _entry: DiscoveryEntry,
    _seq: number,
    intervalMs = 30_000,
  ): void {
    if (this.heartbeatTimers.has(agentId)) {
      clearInterval(this.heartbeatTimers.get(agentId)!);
    }
    const timer = setInterval(
      () => this.heartbeat(agentId).catch(e =>
        console.warn(`[Libp2pDiscovery] heartbeat error for ${agentId}: ${e.message}`)
      ),
      intervalMs,
    );
    this.heartbeatTimers.set(agentId, timer);
  }
}

File: templates/python/discovery/libp2p_discovery.py
"""
Libp2pDiscovery — fully P2P discovery backend for Borgkit.

Architecture:
  Transport   : QUIC (via borgkit-libp2p-sidecar Rust binary)
  Routing     : Kademlia DHT  (/borgkit/kad/1.0.0 — isolated from IPFS)
  Local LAN   : mDNS (optional, default on)
  NAT         : DCUtR hole punching + circuit-relay-v2 fallback
  Identity    : secp256k1 keypair from ANR — same key → same PeerId

WHY A SIDECAR?
  py-libp2p is incomplete as of 2026: it lacks QUIC transport, production-
  grade Kademlia, mDNS, and DCUtR.  Rather than ship a broken pure-Python
  implementation, this module delegates to either:

  1. A compiled `borgkit-libp2p-sidecar` binary (Rust, from templates/rust/)
     — launched as a subprocess, communicates over stdin/stdout JSON-RPC 2.0.

  2. A Node.js process running the TypeScript Libp2pDiscovery
     — used if the sidecar binary is not found but `node` is available.

  3. HttpDiscovery pointed at a local gateway (automatic fallback)
     — used if neither sidecar nor Node is available.
     Set BORGKIT_LIBP2P_GATEWAY=http://localhost:7731 to enable.

  Configure via environment variables or Libp2pDiscoveryConfig:
    BORGKIT_LIBP2P_SIDECAR   path to the compiled sidecar binary
    BORGKIT_LIBP2P_NODE      path to the TypeScript entry point (if using Node)
    BORGKIT_BOOTSTRAP_PEERS  comma-separated multiaddrs of bootstrap peers
    BORGKIT_P2P_PORT         UDP port for QUIC listener (default: 0 = OS-assigned)

Usage:
    cfg = Libp2pDiscoveryConfig(private_key_bytes=my_anr_key)
    discovery = await Libp2pDiscovery.start(cfg)
    await discovery.register(entry)
    peers = await discovery.query('web_search')
    await discovery.stop()
"""

from __future__ import annotations

import asyncio
import json
import os
import shutil
import sys
import dataclasses
from typing import Optional

from interfaces.iagent_discovery import IAgentDiscovery, DiscoveryEntry, NetworkInfo, HealthStatus


# ── Config ────────────────────────────────────────────────────────────────────

@dataclasses.dataclass
class Libp2pDiscoveryConfig:
    """Configuration for the Libp2pDiscovery sidecar."""

    # 32-byte secp256k1 private key (same as ANR signing key).
    # Omit only for ephemeral/throwaway nodes.
    private_key_bytes: Optional[bytes] = None

    # UDP port for the QUIC listener (0 = OS-assigned).
    listen_port: int = 0

    # Bootstrap peer multiaddrs (format: /ip4/.../udp/.../quic-v1/p2p/...)
    # Also read from BORGKIT_BOOTSTRAP_PEERS env var (comma-separated).
    bootstrap_peers: list[str] = dataclasses.field(default_factory=list)

    # How often to re-publish DHT records (seconds). Default: 30
    heartbeat_interval_secs: int = 30

    # Enable mDNS for local network discovery. Default: True
    enable_mdns: bool = True

    # Path to the compiled borgkit-libp2p-sidecar binary.
    # Falls back to BORGKIT_LIBP2P_SIDECAR env var, then PATH lookup.
    sidecar_binary: Optional[str] = None

    # Path to the Node.js TypeScript entry point (fallback if no sidecar).
    # Falls back to BORGKIT_LIBP2P_NODE env var.
    node_entry: Optional[str] = None


# ── JSON-RPC 2.0 helpers ──────────────────────────────────────────────────────

_rpc_id = 0

def _make_request(method: str, params: dict) -> bytes:
    global _rpc_id
    _rpc_id += 1
    msg = {'jsonrpc': '2.0', 'id': _rpc_id, 'method': method, 'params': params}
    return (json.dumps(msg) + '\n').encode()


def _parse_response(line: bytes) -> dict:
    return json.loads(line.decode().strip())


# ── Sidecar process management ────────────────────────────────────────────────

class _SidecarProcess:
    """Manages a borgkit-libp2p-sidecar subprocess (Rust or Node.js)."""

    def __init__(self, proc: asyncio.subprocess.Process):
        self._proc = proc
        self._lock = asyncio.Lock()

    @classmethod
    async def launch(cls, cfg: Libp2pDiscoveryConfig) -> '_SidecarProcess':
        binary = cls._resolve_binary(cfg)
        if binary is None:
            raise RuntimeError(
                'borgkit-libp2p-sidecar not found.\n'
                'Options:\n'
                '  1. Compile the Rust sidecar: cd templates/rust && cargo build --release\n'
                '  2. Set BORGKIT_LIBP2P_SIDECAR=/path/to/borgkit-libp2p-sidecar\n'
                '  3. Set BORGKIT_LIBP2P_GATEWAY=http://localhost:7731 to use HTTP fallback\n'
                '  4. Set BORGKIT_DISCOVERY_URL to use a centralised registry'
            )

        env_peers = os.environ.get('BORGKIT_BOOTSTRAP_PEERS', '')
        all_peers = cfg.bootstrap_peers + [p for p in env_peers.split(',') if p]

        args = [binary, 'run',
                '--port',        str(cfg.listen_port),
                '--heartbeat',   str(cfg.heartbeat_interval_secs),
                '--mdns',        str(cfg.enable_mdns).lower(),
                '--bootstrap',   ','.join(all_peers),
        ]
        if cfg.private_key_bytes:
            args += ['--key', cfg.private_key_bytes.hex()]

        proc = await asyncio.create_subprocess_exec(
            *args,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        return cls(proc)

    @staticmethod
    def _resolve_binary(cfg: Libp2pDiscoveryConfig) -> Optional[str]:
        # 1. Explicit config
        if cfg.sidecar_binary and os.path.isfile(cfg.sidecar_binary):
            return cfg.sidecar_binary
        # 2. Env var
        env_path = os.environ.get('BORGKIT_LIBP2P_SIDECAR')
        if env_path and os.path.isfile(env_path):
            return env_path
        # 3. PATH lookup
        found = shutil.which('borgkit-libp2p-sidecar')
        if found:
            return found
        # 4. Node.js fallback
        node_entry = cfg.node_entry or os.environ.get('BORGKIT_LIBP2P_NODE')
        if node_entry and shutil.which('node'):
            return f'node {node_entry}'  # handled specially in launch
        return None

    async def call(self, method: str, params: dict) -> dict:
        async with self._lock:
            req = _make_request(method, params)
            self._proc.stdin.write(req)
            await self._proc.stdin.drain()
            line = await self._proc.stdout.readline()
            return _parse_response(line)

    async def stop(self) -> None:
        try:
            self._proc.stdin.write((_make_request('stop', {}) + b''))
            await self._proc.stdin.drain()
        except Exception:
            pass
        try:
            self._proc.terminate()
            await asyncio.wait_for(self._proc.wait(), timeout=5.0)
        except Exception:
            self._proc.kill()


# ── HttpDiscovery gateway fallback ────────────────────────────────────────────

class _HttpGatewayFallback(IAgentDiscovery):
    """
    Fallback when the libp2p sidecar is unavailable.
    Connects to a local or remote HTTP gateway that bridges HTTP ↔ libp2p DHT.
    Set BORGKIT_LIBP2P_GATEWAY=http://localhost:7731 to activate.
    """

    def __init__(self, base_url: str):
        from discovery.http_discovery import HttpDiscovery
        self._inner = HttpDiscovery(base_url=base_url)

    async def register(self, entry: DiscoveryEntry) -> None:
        return await self._inner.register(entry)

    async def unregister(self, agent_id: str) -> None:
        return await self._inner.unregister(agent_id)

    async def query(self, capability: str) -> list[DiscoveryEntry]:
        return await self._inner.query(capability)

    async def list_all(self) -> list[DiscoveryEntry]:
        return await self._inner.list_all()

    async def heartbeat(self, agent_id: str) -> None:
        return await self._inner.heartbeat(agent_id)


# ── Libp2pDiscovery ───────────────────────────────────────────────────────────

class Libp2pDiscovery(IAgentDiscovery):
    """
    P2P discovery backend backed by a borgkit-libp2p-sidecar subprocess.

    Falls back to _HttpGatewayFallback if the sidecar is unavailable and
    BORGKIT_LIBP2P_GATEWAY is set.
    """

    def __init__(self, sidecar: '_SidecarProcess'):
        self._sidecar = sidecar

    @classmethod
    async def start(cls, cfg: Libp2pDiscoveryConfig = None) -> 'Libp2pDiscovery':
        if cfg is None:
            cfg = Libp2pDiscoveryConfig()
        try:
            sidecar = await _SidecarProcess.launch(cfg)
            return cls(sidecar)
        except RuntimeError as e:
            gateway = os.environ.get('BORGKIT_LIBP2P_GATEWAY')
            if gateway:
                print(f'[Libp2pDiscovery] Sidecar unavailable — falling back to HTTP gateway: {gateway}')
                return _HttpGatewayFallback(gateway)  # type: ignore[return-value]
            raise RuntimeError(
                f'{e}\n\nAlternatively, set BORGKIT_LIBP2P_GATEWAY=<url> to use an HTTP-to-DHT bridge.'
            ) from e

    async def stop(self) -> None:
        await self._sidecar.stop()
        print('[Libp2pDiscovery] Stopped')

    # ── IAgentDiscovery ───────────────────────────────────────────────────────

    async def register(self, entry: DiscoveryEntry) -> None:
        resp = await self._sidecar.call('register', {'entry': _entry_to_dict(entry)})
        _check_rpc_error(resp)

    async def unregister(self, agent_id: str) -> None:
        resp = await self._sidecar.call('unregister', {'agent_id': agent_id})
        _check_rpc_error(resp)

    async def query(self, capability: str) -> list[DiscoveryEntry]:
        resp = await self._sidecar.call('query', {'capability': capability})
        _check_rpc_error(resp)
        return [_dict_to_entry(d) for d in resp.get('result', [])]

    async def list_all(self) -> list[DiscoveryEntry]:
        resp = await self._sidecar.call('list_all', {})
        _check_rpc_error(resp)
        return [_dict_to_entry(d) for d in resp.get('result', [])]

    async def heartbeat(self, agent_id: str) -> None:
        resp = await self._sidecar.call('heartbeat', {'agent_id': agent_id})
        _check_rpc_error(resp)


# ── Serialisation helpers ─────────────────────────────────────────────────────

def _check_rpc_error(resp: dict) -> None:
    if 'error' in resp:
        raise RuntimeError(f"[Libp2pDiscovery] RPC error: {resp['error']}")


def _entry_to_dict(e: DiscoveryEntry) -> dict:
    return {
        'agentId':      e.agent_id,
        'name':         e.name,
        'owner':        e.owner,
        'capabilities': e.capabilities,
        'network': {
            'protocol': e.network.protocol,
            'host':     e.network.host,
            'port':     e.network.port,
            'tls':      e.network.tls,
        },
        'health': {
            'status':        e.health.status,
            'lastHeartbeat': e.health.last_heartbeat,
            'uptimeSeconds': e.health.uptime_seconds,
        },
        'registeredAt': e.registered_at,
        'metadataUri':  e.metadata_uri,
    }


def _dict_to_entry(d: dict) -> DiscoveryEntry:
    net = d.get('network', {})
    hlt = d.get('health', {})
    return DiscoveryEntry(
        agent_id=d['agentId'],
        name=d['name'],
        owner=d.get('owner', ''),
        capabilities=d.get('capabilities', []),
        network=NetworkInfo(
            protocol=net.get('protocol', 'http'),
            host=net.get('host', ''),
            port=net.get('port', 0),
            tls=net.get('tls', False),
            peer_id=net.get('peerId', ''),
            multiaddr=net.get('multiaddr', ''),
        ),
        health=HealthStatus(
            status=hlt.get('status', 'unknown'),
            last_heartbeat=hlt.get('lastHeartbeat', ''),
            uptime_seconds=hlt.get('uptimeSeconds', 0),
        ),
        registered_at=d.get('registeredAt', ''),
        metadata_uri=d.get('metadataUri'),
    )
